crates/bitwarden/src/state/state.rs

```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    #[cfg(feature = "internal")]
+    pub(crate) async fn set_account_sync_data(
+        &self,
+        id: Uuid,
+        data: bitwarden_api_api::models::SyncResponseModel,
+    ) -> Result<()> {
+        // Before we create the storage profile, keep a copy of the current temporary storage (tokens, kdf params, etc)
+
+        use crate::client::{keys::store_keys_from_sync, profile::store_profile_from_sync};
+        let state = self.account.lock().await.get();
+
+        // Create the new account state, and load the temporary storage into it
+        let mut account = self.account.lock().await;
+        *account = StateStorage::new(id.to_string(), load_medium(&self.path));
+        account
+            .modify(|acc| {
+                *acc = Some(state);
+                Ok(())
+            })
+            .await?;
+        drop(account);
+
+        // Save the new data
+        let profile = data.profile.ok_or(Error::MissingFields)?;
+
+        store_keys_from_sync(profile.as_ref(), self).await?;
+        store_profile_from_sync(profile.as_ref(), self).await?;
+
+        Ok(())
+    }
+
+    #[cfg(feature = "internal")]
+    pub(crate) async fn load_account(&self, id: Uuid) -> Result<()> {
+        let mut account = self.account.lock().await;
+        *account = StateStorage::new(id.to_string(), load_medium(&self.path));
+        account.read().await?;
+        Ok(())
+    }
+}
+
+pub struct StateStorage {
+    cache: HashMap<String, Value>,
+    account_id: String,
+    medium: Box<dyn StateStorageMedium>,
+}
+
+impl StateStorage {
+    fn new(account_id: String, medium: Box<dyn StateStorageMedium>) -> Self {
+        Self {
+            cache: HashMap::new(),
+            account_id,
+            medium,
+        }
+    }
+
+    pub fn get(&self) -> HashMap<String, Value> {
+        self.cache.clone()
+    }
+
+    #[cfg(feature = "internal")]
+    pub async fn read(&mut self) -> Result<HashMap<String, Value>> {
+        let value = self
+            .medium
+            .read()
+            .await?
+            .remove(&self.account_id)
+            .unwrap_or_default();
+        self.cache = value.clone();
+        Ok(value)
+    }
+
+    pub async fn modify<'b>(
+        &mut self,
+        modify_fn: impl FnOnce(&mut Option<HashMap<String, Value>>) -> Result<()> + Send + 'b,
+    ) -> Result<()> {
+        let account_id = self.account_id.clone();
+
+        let result = self
+            .medium
+            .modify(Box::new(move |state| {
+                let mut map = state.remove(&account_id);
+                modify_fn(&mut map)?;
+                if let Some(map) = map {
+                    state.insert(account_id, map);
+                }
+
+                Ok(())
+            }))
+            .await?;
+
+        self.cache = result.get(&self.account_id).cloned().unwrap_or_default();
+        Ok(())
+    }
+}
+
+fn load_medium(path: &Option<String>) -> Box<dyn StateStorageMedium> {
```

Comment:
Is medium some terminology?

Reply:
I've used storage medium to refer to the different storage backends, in a similar way than CD's or DVD's would be different storage media.

It doesn't really add anything to the name though, we can just rename the `StateStorageMedium` trait to `Storage` plainly, and rename `load_medium` to `get_storage` (Because it's not actually loading anything yet).